src/routes/blog/post/remix-3-whats-changing-and-why-it-matters/+page.markdoc

```diff
@@ -0,0 +1,144 @@
+
+For example, a comments section could live inside a Frame. When a user submits a new comment, the server renders updated HTML for that frame and sends it down. The rest of the page stays untouched. It feels like client-side interactivity, but it's still server-driven and progressively enhanced.
+
+### Data loading and actions
+
+The loader and action pattern that defined Remix 2 is still here. Each route can export handlers for different HTTP methods. A `GET` handler fetches data to render; a `POST` (or `PUT`, `DELETE`) handler performs a mutation.
+
+Here's a simple example:
+
+```ts
+export const handlers = {
+  async GET({ params }) {
+    const post = await getPost(params.slug);
+    if (!post) return render(<NotFound />, { status: 404 });
+    return render(<PostPage post={post} />);
+  },
+
+  async POST({ request, params }) {
+    const form = await request.formData();
+    const comment = form.get("comment");
+    await saveComment(params.slug, comment);
+    return redirect(routes.posts.show.href({ slug: params.slug }));
+  }
+};
+```
+
+In this setup, the `GET` handler runs on the server and renders HTML for the initial page. The `POST` handler processes a submitted `<form>`, updates data, and redirects. Remix then re-runs the necessary loaders to refresh the UI. If JavaScript is disabled, the browser still handles the POST and redirect normally, since everything uses standard HTML forms.
+
+This pattern keeps the app predictable. Data always comes from one place, the loader for that route, and every action leads to a clear revalidation path. No client-side fetching hooks or state libraries are needed.
+
+### Error handling
+
+Error handling in Remix 3 continues the route-based model from Remix 2. Each route can define its own error boundary so that failures stay local. If a child route fails to load or render, only that section shows an error message, while the rest of the page remains intact.
+
+When a loader throws a `Response` with an error status, Remix automatically triggers the matching error boundary. For example, if a loader throws `new Response("Not found", { status: 404 })`, Remix will show the Not Found UI for that route but keep parent layouts mounted.
+
+This approach prevents a single broken component from blanking the entire page (a common problem in many SPAs) and helps developers handle network or data errors cleanly.
+
+### TypeScript and developer experience
+
+Remix 3 is written in TypeScript from the start, so every route, loader, and action is strongly typed. Route parameters and return types are inferred automatically. For example, if `routes.books.show` has a `:slug` parameter, TypeScript will know that `params.slug` is a string inside its handlers.
```

Comment:
Do they mention anything about schema validaiton like Zod? Might be nice to mention here.